### PR TITLE
feat(csprng): implement AppleSecureEnclaveSeeder

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves

closees https://github.com/zama-ai/concrete-core-internal/issues/207

### Description

- re-use work from this original PR: https://github.com/zama-ai/concrete-core/pull/4

C FFI to follow

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Review by agnesLeroy — COMMENTED** [review]
We need to wrap this seeder into concrete-core as well no?

**Comment by IceTDrinker** [comment]
Hmm, normally it's not required, we take a Box<dyn Seeder>, I will need to write the FFI though

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        if split_string.len() >= 2 {
+            minor = match split_string[1].parse() {
+                Ok(minor_from_str) => minor_from_str,
+                _ => return false,
+            };
+        }
+
+        // SecRandomCopyBytes is available starting with mac OS 10.7
+        // https://developer.apple.com/documentation/security/1399291-secrandomcopybytes?language=objc
+        major > 10 || major == 10 && minor >= 7
```
Comment (tmontaigu) [inline concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:119]: Maybe we can slightly reduce this by doing 
```rust
let split_string: Vec<&str> = result_string.split('.').collect();

if split_string.len() != 2 {
    return false;
}

match (split_string[0].parse::<u32>(), split_string[1].parse::<u32>()) {
    (Ok(major), Ok(minor)) => major > 10 || major == 10 && minor >= 7,
    _ => false
}
```
or 

```rust
let numbers: Result<Vec<u32>, _> = result_string
    .split('.')
    .map(str::parse)
    .collect();
    
match numbers {
    Ok(nums) if nums.len() == 2 => {
        let major = nums[0];
        let minor = nums[1];

        major > 10 || major == 10 && minor >= 7
    },
    _ => false
}
```

:thinking:

File: concrete-csprng/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,12 @@ keywords = ["fully", "homomorphic", "encryption", "fhe", "cryptography"]
 aes-soft = "0.6.4"
 rayon = { version = "1.5.0", optional = true }
 
+[target.aarch64-apple-darwin.dependencies]
+libc = "0.2.133"
+
+[target.x86_64-apple-darwin.dependencies]
+libc = "0.2.133"
+
```
Comment (tmontaigu) [inline concrete-csprng/Cargo.toml:23]: I think we should be able to do

```toml
[target.'cfg(target_os = "macos")'.dependencies]
libc = "0.2.133"
```

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+/// <https://www.felixcloutier.com/x86/rdseed> .
+pub struct AppleSecureEnclaveSeeder;
+
+impl Seeder for AppleSecureEnclaveSeeder {
+    fn seed(&mut self) -> Seed {
+        // 16 bytes == 128 bits
+        let mut bytes = [0u8; 16];
+        if let Err(err) = secure_enclave::generate_random_bytes(&mut bytes) {
+            panic!("Failure while using Apple secure enclave: {err:?}")
+        }
```
Comment (tmontaigu) [inline concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:59]: We may as well use `expect`:

```rust
 secure_enclave::generate_random_bytes(&mut bytes)
   .expect("Failure while using Apple secure enclave");

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:119]: I went for the overly dumb version not knowing if the format would always be major.minor and not major.minor.patch down the line

File: concrete-csprng/Cargo.toml (hunk 2 above)
Comment (IceTDrinker) [reply concrete-csprng/Cargo.toml:23]: Yeah it’s the resolver 2 debate which I haven’t settled in my mind yet, plus the way it is handled currently departs from using features -_- but it looks so annoying to have a mac OS specific feature

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:59]: Yes, but I want the underlying error to be displayed, I though expect dismissed the original error altogether ?

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:119]: Here also it should handle a case where the version format is only the major version if I'm not mistaken

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:59]: Hmm... I don't know how I managed to misread expect documentation 🤔 it says it does print the underlying error thanks !

File: concrete-csprng/Cargo.toml (hunk 2 above)
Comment (tmontaigu) [reply concrete-csprng/Cargo.toml:23]: I don't think we need resolver to for that particular use no ?

Comment (IceTDrinker) [reply concrete-csprng/Cargo.toml:23]: it seems to be ok indeed

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs]: updated the logic

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-csprng/src/main.rs [hunk 4]
```diff
@@ -1,13 +1,27 @@
 //! This program uses the concrete csprng to generate an infinite stream of random bytes on
 //! the program stdout. For testing purpose.
-use concrete_csprng::generators::{AesniRandomGenerator, RandomGenerator};
-use concrete_csprng::seeders::{RdseedSeeder, Seeder};
+#[cfg(target_arch = "x86_64")]
+use concrete_csprng::generators::AesniRandomGenerator as ActivatedRandomGenerator;
+#[cfg(target_arch = "aarch64")]
+use concrete_csprng::generators::ArmAesRandomGenerator as ActivatedRandomGenerator;
```
Comment (agnesLeroy) [inline concrete-csprng/src/main.rs:6]: We call it ArmAes? NeonAes would maybe be more accurate?

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            if res != 0 {
+                Err(std::io::Error::last_os_error())
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+/// A seeder which uses the `rdseed` x86_64 instruction.
```
Comment (agnesLeroy) [inline concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:48]: ARM Neon aes instruction set

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+///
+/// The `rdseed` instruction allows to deliver seeds from a hardware source of entropy see
```
Comment (agnesLeroy) [inline concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:50]: Here again rdseed -> Arm neon aes

**Review by agnesLeroy — COMMENTED** [review]
Hello! Thanks a lot for this! I just have a naming question you'll see.

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 6 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:50]: this is independent from Arm neon and aes but thanks for pointing it out !

File: concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs (hunk 5 above)
Comment (IceTDrinker) [reply concrete-csprng/src/seeders/implem/apple_secure_enclave_seeder.rs:48]: this is independent from Arm neon and aes but thanks for pointing it out !

File: concrete-csprng/src/main.rs (hunk 4 above)
Comment (IceTDrinker) [reply concrete-csprng/src/main.rs:6]: We use aesni for the x86 version which apparently is the name of the instruction in the x86 world, here just calling it AesRandomGenerator would have been weird so I added the name of the platform ArmAesRandomGenerator

Not sure about adding neon in there :/

Comment (IceTDrinker) [reply concrete-csprng/src/main.rs:6]: Fine for naming it `NeonAesRandomGenerator` will make the change

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Review by tmontaigu — APPROVED** [review]

**Review by agnesLeroy — APPROVED** [review]

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Pull request merged**